### PR TITLE
Thread safe entity registry

### PR DESCRIPTION
**Pull request by TheoBerlin**
Creating a component or entity leads to writes in the entity registry. All reads and writes in the entity registry should be thread locked to avoid data races.

fixes #76

### CODE REVIEW

File: LambdaEngine/Source/ECS/EntityRegistry.cpp [hunk 1]
```diff
@@ -1,90 +1,97 @@
...
+		std::scoped_lock<SpinLock> lock(m_Lock);
+
+		EntityRegistryPage& topPage = m_EntityPages.top();
+		if (!topPage.HasElement(entity)) {
+			// Initialize a new set
+			topPage.PushBack({ componentType }, entity);
+		} else {
+			// Add the component type to the set
+			topPage.IndexID(entity).insert(componentType);
+		}
```
Comment (Vonsen) [inline LambdaEngine/Source/ECS/EntityRegistry.cpp]: curly brackets

File: LambdaEngine/Source/ECS/EntityRegistry.cpp [hunk 2]
```diff
@@ -1,90 +1,97 @@
...
+	void EntityRegistry::DeregisterComponentType(Entity entity, std::type_index componentType)
+	{
+		std::scoped_lock<SpinLock> lock(m_Lock);
+
+		EntityRegistryPage& topPage = m_EntityPages.top();
+		if (!topPage.HasElement(entity)) {
+			LOG_WARNING("Attempted to deregister a component type (%s) from an unregistered entity: %ld", componentType.name(), entity);
+		} else {
+			topPage.IndexID(entity).erase(componentType);
+		}
```
Comment (Vonsen) [inline LambdaEngine/Source/ECS/EntityRegistry.cpp]: curly brackets

File: LambdaEngine/Source/ECS/EntityRegistry.cpp [hunk 3]
```diff
@@ -1,90 +1,97 @@
...
+
+		const EntityRegistryPage& topPage = m_EntityPages.top();
+		const std::unordered_set<std::type_index>& entityTypes = topPage.IndexID(entity);
+
+		for (std::type_index type : queryTypes) {
+			auto got = entityTypes.find(type);
+			if (got == entityTypes.end()) {
+				return false;
+			}
+		}
```
Comment (Vonsen) [inline LambdaEngine/Source/ECS/EntityRegistry.cpp]: curly bracket

File: LambdaEngine/Source/ECS/EntityRegistry.cpp [hunk 4]
```diff
@@ -1,90 +1,97 @@
...
+
+	void EntityRegistry::DeregisterEntity(Entity entity)
+	{
+		std::scoped_lock<SpinLock> lock(m_Lock);
+
+		EntityRegistryPage& topPage = m_EntityPages.top();
+		if (!topPage.HasElement(entity)) {
+			LOG_WARNING("Attempted to deregister an unregistered entity: %ld", entity);
+			return;
+		}
```
Comment (Vonsen) [inline LambdaEngine/Source/ECS/EntityRegistry.cpp]: curly brackets

File: LambdaEngine/Source/ECS/EntityRegistry.cpp [hunk 5]
```diff
@@ -1,90 +1,97 @@
...
+		m_EntityPages.push({});
+	}
+
+	void EntityRegistry::RemovePage()
+	{
+		EntityRegistryPage& topPage = m_EntityPages.top();
+		const TArray<Entity> entities = topPage.GetIDs();
+		for (Entity entity : entities) {
+			m_EntityIDGen.PopID(entity);
+		}
```
Comment (Vonsen) [inline LambdaEngine/Source/ECS/EntityRegistry.cpp]: curly bracket

**Review by Vonsen — CHANGES_REQUESTED** [review]
fix curly brackets and lint warning

File: LambdaEngine/Source/ECS/EntityRegistry.cpp (hunk 1 above)
Comment (TheoBerlin) [reply LambdaEngine/Source/ECS/EntityRegistry.cpp]: Done

**Review by Vonsen — APPROVED** [review]

**Pull request merged**